ResearchKit/Common/ORKFormStepViewController.m

```diff
@@ -692,6 +746,10 @@ - (ORKStepResult *)result {
     
     NSMutableArray *qResults = [NSMutableArray new];
     for (ORKFormItem *item in items) {
+        
+        if ([_filteredForms containsObject:item]) {
```

Comment:
Is it necessary to add an ivar for the filtered forms vs evaluating the predicate again?

Reply:
I originally wanted to do it that way, but to evaluate the predicate, you need a current version of what this function outputs which then leads to a circular dependency :(

Reply:
Interesting, if thats the case it seems like the circular dependency is still there and we are performing a single cycle every time `filterSections:` is called? Is there a way to get the full results with no filtering, and pass that into the predicate? Otherwise I feel like you could have ambiguous predicates where item A depends on item B, and item B depends on item A and it's unclear which takes precedent.

Reply:
It is a tricky "chicken or egg" scenario. But I think they are not dependent upon each other because when `filterSections:` gets called, it's emptying out the `_filteredForms` array before requesting the current `result` so that logic technically won't rely on the predicate at that time. However, when navigating into the next step, result will be called via navigation methods (more specifically via `[ORKTaskViewController stepViewController:didFinishWithNavigationDirection:]`). In that case the `_filteredForms` will be populated with the latest state at completion of the FormStep and any filtered results will be removed before passing up the hierarchy. 
I still think there's a possibility to have interfering predicates like you suggest. Without have a validation system at construction, I'm not sure how to solve that problem.

Reply:
Ah I see what you're saying. That makes sense. LGTM